### PR TITLE
DATACOUCH-275 Key auto generation using Attributes, UUID

### PR DESCRIPTION
**Pull request by subalakr**

### CODE REVIEW

File: src/main/asciidoc/autokeygeneration.adoc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+NOTE: The maximum key length supported by couchbase is 250 bytes.
+
+[[couchbase.autokeygeneration.configuration]]
+== Configuration
+
+Keys to be auto-generated should be annotated with `@GeneratedValue`. The default strategy is `USE_ATTRIBUTES`. Prefix
+and suffix for the key can be provided as part of the entity itself, these values are not persisted, they are only
+used for key generation. The prefixes and suffixes are ordered using the `order` value. The default order is `0`, multiple
+prefixes without order will overwrite the previous. If a value for id is already available, auto-generation will be skipped.
+The deimiter for concatentation can be provided using `delimiter`, the default delimiter is `.`.
```
Comment (simonbasle) [inline src/main/asciidoc/autokeygeneration.adoc]: typo: `deimiter` instead of `delimiter`

File: src/main/asciidoc/autokeygeneration.adoc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+String id = couchbaseTemplate.getGeneratedId(entity);
+...
+repo.exists(id);
+----
+====
+
+[[couchbase.autokeygeneration.usingattributes]]
+== Key generation using attributes
+
+It is a common practise to generate keys using a combination of the document attributes. Key generation using attributes
```
Comment (simonbasle) [inline src/main/asciidoc/autokeygeneration.adoc]: typo: `practise` -> `practice`

File: src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java [hunk 3]
```diff
@@ -836,4 +889,49 @@ public ConverterAwareSpELExpressionParameterValueProvider(final SpELExpressionEv
     }
   }
 
+  private String generateId(GeneratedValue generatedValue, TreeMap<Integer, String> prefixes, TreeMap<Integer, String> suffixes,
+                            TreeMap<Integer, String> idAttributes) {
+    String delimiter = generatedValue.delimiter();
+    StringBuilder sb = new StringBuilder();
+    boolean isAppending = false;
+    if (prefixes.size() > 0) {
+        appendKeyParts(sb, prefixes.values(), generatedValue.delimiter());
```
Comment (simonbasle) [inline src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java]: can reuse the `delimiter` variable here (applies also to the blocks below)

File: src/main/java/org/springframework/data/couchbase/core/mapping/KeySettings.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	 */
+	public KeySettings suffix(String suffix) {
+		this.commonSuffix = suffix;
+		return this;
+	}
+
+	/**
+	 * Set delimiter
+	 */
+	public KeySettings delimiter(String delimiter) {
```
Comment (simonbasle) [inline src/main/java/org/springframework/data/couchbase/core/mapping/KeySettings.java]: should probably test for `null` delimiter and either ignore or preferably interpret null as "no delimiter" (the empty `String`)

File: src/main/java/org/springframework/data/couchbase/core/mapping/id/GenerationStrategy.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+public enum GenerationStrategy {
+	/**
+	 * Constructs key from the entity attributes using the supplied id, prefix, suffix
+	 */
+	USE_ATTRIBUTES,
+
+	/**
+	 * Uses an uuid generator
+	 */
+	UNIQUE
```
Comment (simonbasle) [inline src/main/java/org/springframework/data/couchbase/core/mapping/id/GenerationStrategy.java]: have you guys thought about another "unique" strategy based on a counter document?

Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/id/GenerationStrategy.java]: Yes thought about it for sequential id generation, but it is a backlog right now as there will be a performance impact of a hidden counter operation each time.

File: src/main/java/org/springframework/data/couchbase/core/mapping/KeySettings.java (hunk 4 above)
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/KeySettings.java]: Yes, will add null check

File: src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java (hunk 3 above)
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java]: sure, will update it.

File: src/main/asciidoc/autokeygeneration.adoc (hunk 2 above)
Comment (subalakr) [inline src/main/asciidoc/autokeygeneration.adoc]: Thanks, will fix.

File: src/main/asciidoc/autokeygeneration.adoc (hunk 1 above)
Comment (subalakr) [inline src/main/asciidoc/autokeygeneration.adoc]: Thanks, will fix.

**Comment by daschl** [comment]
Looks good to me.

**Comment by subalakr** [comment]
Merged

**Pull request closed without merging**